Simplify trim-paths feature by merging all debuginfo options together

Urgau opened the pull request:
This PR simplifies the trim-paths feature by merging all debuginfo options together, as described in https://github.com/rust-lang/rust/issues/111540#issuecomment-1994010274.

And also do some correctness fixes found during the review.

cc @weihanglo 
r? @michaelwoerister

michaelwoerister commented:
Thanks for the PR, @Urgau! I think we'll do this, but since this is deviating from the accepted RFC, I want to find out what the procedure is here.

michaelwoerister commented on compiler/rustc_codegen_llvm/src/back/write.rs:
```diff
@@ -257,13 +257,12 @@ pub fn target_machine_factory(
         let path_to_cstring_helper = |path: Option<PathBuf>| -> CString {
             let path = path.unwrap_or_default();
-            let path = if should_prefer_remapped_for_split_debuginfo_paths {
+            let path = if should_prefer_remapped_paths {
```
I can't quite figure out if this path is emitted into something or is just the path where LLVM will actually put the file on disk. Or both 🙂

We'd only want to remap if it is written into debuginfo.

Urgau replied:
I think to remember that it is emitted into debuginfo, or at least it was since this place already had the remapping in place before I updated it for the trim-path feature.

michaelwoerister replied:
Yep, you're right. This is the path that ends up in the `DW_AT_GNU_dwo_name` field of the compilation unit DIE.

Urgau commented:
> I think we'll do this, but since this is deviating from the accepted RFC, I want to find out what the procedure is here.

I'm not aware of such procedure. Our RFCs are not normative and it's common to diverged from them in the implementation and experimentation phase.

I would say that as long as we respect the goal of the RFC (as defined in the Summary and Motivation section of the RFC), we are good to adapt the implementation to achieve that goal, particularly here since we are adapting the implementation to match the reality (ie. the current limitation of the tooling and LLVM).

There is also nothing here that would prevent re-introducing the lost options in the future.

So I think we are go to go.

michaelwoerister commented:
I actually remembered a similar case where I asked about the proceduce:
https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Add.20-C.20dlltool.20for.20raw-dylibs.3F

Let's make sure we document what we are doing on the tracking issue an mention it in the stabilization FCP, then we should indeed be good.

michaelwoerister commented on compiler/rustc_session/src/session.rs:
```diff
@@ -899,37 +899,7 @@ impl Session {
-            .remap_path_scope
-            .contains(RemapPathScopeComponents::SPLIT_DEBUGINFO_PATH)
-            && has_split_debuginfo
+        self.opts.unstable_opts.remap_path_scope.contains(RemapPathScopeComponents::DEBUGINFO)
```
I think the `for_codegen` terminology is a bit problematic. Originally, before scopes and split-debuginfo were a thing, it meant "anything that ends up in a file". But now it is not immediately clear if this should cover the MACRO scope too.

Can we replace this method with `fn should_prefer_remapped(&self, scope: RemapPathScopeComponents) -> bool`?

That might also help uncover bugs like with `Session::local_crate_source_file` that is also used in some diagnostic contexts.

Urgau replied:
Sure. Pushed a second commit doing the removal.

michaelwoerister replied:
Thanks!

michaelwoerister commented on compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs:
```diff
@@ -174,8 +174,14 @@ impl GlobalFileTable {
+            .sess
+            .opts
+            .working_dir
+            .for_scope(tcx.sess, RemapPathScopeComponents::DEBUGINFO)
```
@Zalathar, I found you in triagebot.toml for coverage related questions:

Do you know where the file path generated here will end up in? Constant data? A coverage-specific file? Debuginfo?

Zalathar replied:
It ends up inside a special `__llvm_covmap` section in the final binary. LLVM's `llvm-cov` tool then reads it back out of the binary when preparing coverage reports, in order to find the corresponding source files.

michaelwoerister replied:
Thanks for the info, @Zalathar!

I looks like Clang has its own remapping scope for this via [-fcoverage-prefix-map](https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-fcoverage-prefix-map). We don't have that so I think MACRO is the best fit for now.

Zalathar replied:
Also note that this path ends up in the same place as the paths created in `compiler/rustc_mir_transform/src/coverage/mod.rs`, so whatever the decision is, both should get the same treatment.

Zalathar replied:
I'm not familiar with path remapping, but it does seem very unintuitive to me that this information would be controlled by something named `macro`.

michaelwoerister replied:
Yes, I agree. It's a [known issue](https://github.com/rust-lang/rfcs/pull/3127/files#r882881164) that the name is not a good fit for everything that it covers. I definitely think that this should be fixed before stabilization.

Urgau replied:
Done, switch to `MACRO`.

Btw, I like your suggestion of renaming it to `expansion`.

michaelwoerister commented on compiler/rustc_metadata/src/rmeta/encoder.rs:
```diff
@@ -549,7 +549,11 @@ impl<'a, 'tcx> EncodeContext<'a, 'tcx> {
+                    let adapted_file_name = if self
+                        .tcx
+                        .sess
+                        .should_prefer_remapped(RemapPathScopeComponents::DEBUGINFO)
```
This is kind of an unclear case to me for a number of reasons:

- Crate metadata is not really "debuginfo". One could expect it to be covered by the "object" scope.
- We don't know for certain how the source file paths stored here will be used by downstream compilation sessions. There's nothing preventing the compiler from emitting them to something in the macro or diagnostics scope. 

I think we should err on the side of caution and remap unconditionally (maybe revisit the question later?)

Urgau replied:
Sure. Done.

michaelwoerister commented on compiler/rustc_session/src/session.rs:
```diff
@@ -252,7 +252,8 @@ impl Session {
         let path = self.io.input.opt_path()?;
-        if self.should_prefer_remapped_for_codegen() {
+        // FIXME: The remap path scope should probably not be hardcoded.
+        if self.should_prefer_remapped(RemapPathScopeComponents::DEBUGINFO) {
```
Let's fix the FIXME 🙂 The `local_crate_source_file` method is called in at least one DIAGNOSTIC context.

Urgau replied:
You're right. I've added a new commit switching fixing it.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8331943817/job/22799932045) [(plain)](https://github.com/rust-lang/rust/commit/e4a6f48e963dddd968e417a4f2aa3a300c2b8825/checks/22799932045/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=Urgau
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_0f72c1e3-4c28-4ebd-8b71-03d3b72ed9a0
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=simplify-trim-paths-feature
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_0f72c1e3-4c28-4ebd-8b71-03d3b72ed9a0
GITHUB_REF=refs/pull/122450/merge
GITHUB_REF_NAME=122450/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=5e3814354e7fd663d2545c3d2fbe3994dbedae60
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_0f72c1e3-4c28-4ebd-8b71-03d3b72ed9a0
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_0f72c1e3-4c28-4ebd-8b71-03d3b72ed9a0
GITHUB_TRIGGERING_ACTOR=Urgau
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/122450/merge
GITHUB_WORKFLOW_SHA=5e3814354e7fd663d2545c3d2fbe3994dbedae60
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_20_X64=/opt/hostedtoolcache/go/1.20.14/x64
---
#17 exporting to docker image format
#17 sending tarball 43.5s done
#17 DONE 53.1s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Mon Mar 18 18:41:59 UTC 2024
  network time: Mon, 18 Mar 2024 18:41:59 GMT
  network time: Mon, 18 Mar 2024 18:41:59 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'build.optimized-compiler-builtins', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished release [optimized] target(s) in 3.48s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
---- [run-make] tests/run-make/remap-path-prefix stdout ----

error: make failed
status: exit status: 2
command: cd "/checkout/tests/run-make/remap-path-prefix" && env -u CARGO_MAKEFLAGS -u MAKEFLAGS -u MFLAGS -u RUSTFLAGS AR="ar" CC="cc -ffunction-sections -fdata-sections -fPIC -m64" CXX="c++ -ffunction-sections -fdata-sections -fPIC -m64" HOST_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" LD_LIB_PATH_ENVVAR="LD_LIBRARY_PATH" LLVM_BIN_DIR="/usr/lib/llvm-16/bin" LLVM_COMPONENTS="aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgputargetmca amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfologicalview debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker dwarflinkerparallel dwp engine executionengine extensions filecheck frontendhlsl frontendopenacc frontendopenmp fuzzercli fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irprinter irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker loongarch loongarchasmparser loongarchcodegen loongarchdesc loongarchdisassembler loongarchinfo lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts objcopy object objectyaml option orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvtargetmca runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target targetparser textapi transformutils ve veasmparser vecodegen vectorize vedesc vedisassembler veinfo webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsdriver windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86targetmca xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" LLVM_FILECHECK="/usr/lib/llvm-16/bin/FileCheck" NODE="/usr/bin/node" PYTHON="/usr/bin/python3" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTDOC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" RUST_BUILD_STAGE="stage2-x86_64-unknown-linux-gnu" S="/checkout" TARGET="x86_64-unknown-linux-gnu" TARGET_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" TMPDIR="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/remap-path-prefix/remap-path-prefix" "make"
--- stdout -------------------------------
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/remap-path-prefix/remap-path-prefix:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/remap-path-prefix/remap-path-prefix -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/remap-path-prefix/remap-path-prefix  -Ainternal_features --remap-path-prefix $PWD/auxiliary=/the/aux --crate-type=lib --emit=metadata auxiliary/lib.rs
grep "/the/aux/lib.rs" /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/remap-path-prefix/remap-path-prefix/liblib.rmeta || exit 1
! grep "$PWD/auxiliary" /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/remap-path-prefix/remap-path-prefix/liblib.rmeta || exit 1
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/remap-path-prefix/remap-path-prefix:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/remap-path-prefix/remap-path-prefix -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/remap-path-prefix/remap-path-prefix  -Ainternal_features --remap-path-prefix $PWD/auxiliary=/the/aux -Zremap-path-scope=object  --crate-type=lib --emit=metadata auxiliary/lib.rs
Build completed unsuccessfully in 0:38:16
grep "/the/aux/lib.rs" /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/remap-path-prefix/remap-path-prefix/liblib.rmeta || exit 1
! grep "$PWD/auxiliary" /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/remap-path-prefix/remap-path-prefix/liblib.rmeta || exit 1
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/remap-path-prefix/remap-path-prefix:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/remap-path-prefix/remap-path-prefix -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/remap-path-prefix/remap-path-prefix  -Ainternal_features --remap-path-prefix $PWD/auxiliary=/the/aux -Zremap-path-scope=diagnostics  --crate-type=lib --emit=metadata auxiliary/lib.rs
! grep "/the/aux/lib.rs" /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/remap-path-prefix/remap-path-prefix/liblib.rmeta || exit 1
--- stderr -------------------------------
grep: /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/remap-path-prefix/remap-path-prefix/liblib.rmeta: binary file matches
grep: /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/remap-path-prefix/remap-path-prefix/liblib.rmeta: binary file matches
grep: /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/remap-path-prefix/remap-path-prefix/liblib.rmeta: binary file matches
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #122713) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8338837209/job/22819804393) [(plain)](https://github.com/rust-lang/rust/commit/e12220d7a8a78efe6b96427bdf2c920b916550df/checks/22819804393/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=Urgau
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_1cb743d1-b619-4b63-81da-61323090c6c5
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=simplify-trim-paths-feature
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_1cb743d1-b619-4b63-81da-61323090c6c5
GITHUB_REF=refs/pull/122450/merge
GITHUB_REF_NAME=122450/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=b0bb21fe8a8c02f1645b54fb4966a58dffdfdf73
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_1cb743d1-b619-4b63-81da-61323090c6c5
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_1cb743d1-b619-4b63-81da-61323090c6c5
GITHUB_TRIGGERING_ACTOR=Urgau
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/122450/merge
GITHUB_WORKFLOW_SHA=b0bb21fe8a8c02f1645b54fb4966a58dffdfdf73
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_20_X64=/opt/hostedtoolcache/go/1.20.14/x64
---
#16 exporting to docker image format
#16 sending tarball 45.4s done
#16 DONE 55.9s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
##[group]Clock drift check
  local time: Tue Mar 19 07:02:35 UTC 2024
  network time: Tue, 19 Mar 2024 07:02:35 GMT
  network time: Tue, 19 Mar 2024 07:02:35 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
```

</details>

Urgau commented on compiler/rustc_passes/src/entry.rs:
```diff
@@ -6,7 +6,8 @@ use rustc_hir::def_id::{DefId, LocalDefId, CRATE_DEF_ID, LOCAL_CRATE};
 use rustc_middle::ty::TyCtxt;
-use rustc_session::config::{sigpipe, CrateType, EntryFnType};
+use rustc_session::config::{sigpipe, CrateType, EntryFnType, RemapPathScopeComponents};
+use rustc_session::parse::feature_err;
```
```suggestion
```

michaelwoerister commented on compiler/rustc_mir_transform/src/coverage/mod.rs:
```diff
@@ -124,8 +124,14 @@ fn create_mappings<'tcx>(
+    let file_name = Symbol::intern(
+        &source_file
+            .name
+            .for_scope(tcx.sess, RemapPathScopeComponents::DEBUGINFO)
```
This should also be MARCO. See https://github.com/rust-lang/rust/pull/122450#discussion_r1528386619.

Urgau replied:
Hum, I'm pretty sure I did; I must have forgot this in the fight to fix the test.

```suggestion
            .for_scope(tcx.sess, RemapPathScopeComponents::MACRO)
```

michaelwoerister commented on src/librustdoc/html/render/mod.rs:
```diff
@@ -2506,7 +2507,8 @@ fn render_call_locations<W: fmt::Write>(mut w: W, cx: &mut Context<'_>, item: &c
             let source_map = tcx.sess.source_map();
-            let crate_src = tcx.sess.local_crate_source_file()?;
+            let crate_src =
+                tcx.sess.local_crate_source_file(RemapPathScopeComponents::DIAGNOSTICS)?;
```
This is being compared against `RealFileName::LocalPath` below, that is without any remapping. So we should not remap here either. Can you make `local_crate_source_file` return a `RealFileName`? And since we have the session available there, we probably don't need to pass in the scopes.

Urgau replied:
The logic here explicitly takes the remapping into account, always returning the "real-path" would change this, for better or worth. @GuillaumeGomez would rustdoc prefer to always have "real path" or does rustdoc wants to take into account the remapping here?

As for returning `RealFileName` we could probably do that, not sure if it's a good idea since all the caller seems to want to `PathBuf`/`Path`.

michaelwoerister replied:
> The logic here explicitly takes the remapping into account, always returning the "real-path" would change this

Yes, it looks like the other side of the comparison is also coming from the source map and never expects to that no remapping has happened. 

> As for returning RealFileName we could probably do that, not sure if it's a good idea since all the caller seems to want to PathBuf/Path.

Returning `RealFileName` would retain complete information and push the decision of which PathBuf to use to the usage site. I think that's a good way to do it and would be consistent with other places, like this for example:
https://github.com/rust-lang/rust/blob/8579a1856a868033b792c08e5c73a7c9dbe4d273/compiler/rustc_session/src/options.rs#L224

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8342485487/job/22830736460) [(plain)](https://github.com/rust-lang/rust/commit/6e19aceeb2dfe47dc483bad4cee061b10f99d21a/checks/22830736460/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=Urgau
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_70c593d2-c7b9-4699-8810-86ae70a6a4e3
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=simplify-trim-paths-feature
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_70c593d2-c7b9-4699-8810-86ae70a6a4e3
GITHUB_REF=refs/pull/122450/merge
GITHUB_REF_NAME=122450/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=4df9ee30347a73da737eec5576613f4406d88fe3
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_70c593d2-c7b9-4699-8810-86ae70a6a4e3
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_70c593d2-c7b9-4699-8810-86ae70a6a4e3
GITHUB_TRIGGERING_ACTOR=Urgau
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/122450/merge
GITHUB_WORKFLOW_SHA=4df9ee30347a73da737eec5576613f4406d88fe3
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_20_X64=/opt/hostedtoolcache/go/1.20.14/x64
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.433   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.449 Collecting boolean-py==4.0
#13 0.456   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.474 Collecting chardet==5.1.0
---
#13 3.599 Building wheels for collected packages: reuse
#13 3.600   Building wheel for reuse (pyproject.toml): started
#13 3.927   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.928   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 3.928   Stored in directory: /tmp/pip-ephem-wheel-cache-7admm1wj/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 3.930 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 3.952   Attempting uninstall: setuptools
#13 3.952     Found existing installation: setuptools 59.6.0
#13 3.954     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.954     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.954     Can't uninstall 'setuptools'. No files were found to uninstall.
#13 4.615 Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
#13 4.615 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 5.138 Collecting virtualenv
#13 5.193   Downloading virtualenv-20.25.1-py3-none-any.whl (3.8 MB)
#13 5.371      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 21.5 MB/s eta 0:00:00
#13 5.411 Collecting distlib<1,>=0.3.7
#13 5.419   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.465 Collecting filelock<4,>=3.12.2
#13 5.473   Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
#13 5.473   Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
#13 5.507 Collecting platformdirs<5,>=3.9.1
#13 5.515   Downloading platformdirs-4.2.0-py3-none-any.whl (17 kB)
#13 5.601 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.766 Successfully installed distlib-0.3.8 filelock-3.13.1 platformdirs-4.2.0 virtualenv-20.25.1
#13 DONE 5.8s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      174016 kB
DirectMap2M:     6117376 kB
DirectMap1G:    12582912 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/8579a1856a868033b792c08e5c73a7c9dbe4d273/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-8579a1856a868033b792c08e5c73a7c9dbe4d273-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished release [optimized] target(s) in 27.79s
##[endgroup]
fmt check
##[error]Diff in /checkout/compiler/rustc_mir_transform/src/coverage/mod.rs at line 127:
 
     use rustc_session::{config::RemapPathScopeComponents, RemapFileNameExt};
-        &source_file
-            .name
-            .name
-            .for_scope(tcx.sess, RemapPathScopeComponents::MACRO)
-            .to_string_lossy(),
+        &source_file.name.for_scope(tcx.sess, RemapPathScopeComponents::MACRO).to_string_lossy(),
 
 
     let term_for_bcb = |bcb| {
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_mir_transform/src/coverage/counters.rs" "/checkout/compiler/rustc_mir_transform/src/coverage/mod.rs" "/checkout/compiler/rustc_mir_transform/src/coverage/graph.rs" "/checkout/compiler/rustc_mir_transform/src/coverage/spans/from_mir.rs" "/checkout/compiler/rustc_mir_transform/src/coverage/query.rs" "/checkout/compiler/rustc_mir_transform/src/coverage/spans.rs" "/checkout/compiler/rustc_mir_transform/src/coroutine/by_move_body.rs" "/checkout/compiler/rustc_mir_transform/src/large_enums.rs" "/checkout/compiler/rustc_mir_transform/src/add_call_guards.rs" "/checkout/compiler/rustc_mir_transform/src/remove_zsts.rs" "/checkout/compiler/rustc_mir_transform/src/ssa.rs" "/checkout/compiler/rustc_mir_transform/src/const_debuginfo.rs" "/checkout/compiler/rustc_mir_transform/src/elaborate_box_derefs.rs" "/checkout/compiler/rustc_mir_transform/src/add_retag.rs" "/checkout/compiler/rustc_mir_transform/src/pass_manager.rs" "/checkout/compiler/rustc_mir_transform/src/simplify_branches.rs" "/checkout/compiler/rustc_mir_transform/src/ctfe_limit.rs" "/checkout/compiler/rustc_mir_transform/src/lint.rs" "/checkout/compiler/rustc_mir_transform/src/deduplicate_blocks.rs" "/checkout/compiler/rustc_mir_transform/src/remove_storage_markers.rs" "/checkout/compiler/rustc_mir_transform/src/remove_unneeded_drops.rs" "/checkout/compiler/rustc_mir_transform/src/remove_uninit_drops.rs" "/checkout/compiler/rustc_mir_transform/src/unreachable_prop.rs" "/checkout/compiler/rustc_mir_transform/src/inline/cycle.rs" "/checkout/compiler/rustc_mir_transform/src/jump_threading.rs" "/checkout/compiler/rustc_mir_transform/src/multiple_return_terminators.rs" "/checkout/compiler/rustc_mir_transform/src/deduce_param_attrs.rs" "/checkout/compiler/rustc_mir_transform/src/dataflow_const_prop.rs" "/checkout/compiler/rustc_mir_transform/src/cross_crate_inline.rs" "/checkout/compiler/rustc_mir_transform/src/required_consts.rs" "/checkout/compiler/rustc_mir_transform/src/errors.rs" "/checkout/compiler/rustc_mir_transform/src/simplify_comparison_integral.rs" "/checkout/compiler/rustc_mir_transform/src/nrvo.rs" "/checkout/compiler/rustc_mir_transform/src/coroutine.rs" "/checkout/compiler/rustc_mir_transform/src/simplify.rs" "/checkout/compiler/rustc_mir_transform/src/reveal_all.rs" "/checkout/compiler/rustc_mir_transform/src/add_moves_for_packed_drops.rs" "/checkout/compiler/rustc_mir_transform/src/normalize_array_len.rs" "/checkout/compiler/rustc_mir_transform/src/gvn.rs" "/checkout/compiler/rustc_mir_transform/src/shim.rs" "/checkout/compiler/rustc_mir_transform/src/known_panics_lint.rs" "/checkout/compiler/rustc_mir_transform/src/dest_prop.rs" "/checkout/compiler/rustc_mir_transform/src/lib.rs" "/checkout/compiler/rustc_mir_transform/src/check_unsafety.rs" "/checkout/compiler/rustc_mir_transform/src/ffi_unwind_calls.rs" "/checkout/compiler/rustc_mir_transform/src/dead_store_elimination.rs" "/checkout/compiler/rustc_mir_transform/src/check_const_item_mutation.rs" "/checkout/compiler/rustc_mir_transform/src/inline.rs" "/checkout/compiler/rustc_mir_transform/src/dump_mir.rs" "/checkout/compiler/rustc_mir_transform/src/remove_noop_landing_pads.rs" "/checkout/compiler/rustc_mir_transform/src/copy_prop.rs" "/checkout/compiler/rustc_mir_transform/src/cost_checker.rs" "/checkout/compiler/rustc_mir_transform/src/lower_intrinsics.rs" "/checkout/compiler/rustc_mir_transform/src/ref_prop.rs" "/checkout/compiler/rustc_mir_transform/src/elaborate_drops.rs" "/checkout/compiler/rustc_mir_transform/src/early_otherwise_branch.rs" "/checkout/compiler/rustc_mir_transform/src/deref_separator.rs" "/checkout/compiler/rustc_mir_transform/src/lower_slice_len.rs" "/checkout/compiler/rustc_mir_transform/src/sroa.rs" "/checkout/compiler/rustc_mir_transform/src/remove_place_mention.rs" "/checkout/compiler/rustc_mir_transform/src/function_item_references.rs" "/checkout/compiler/rustc_mir_transform/src/abort_unwinding_calls.rs" "/checkout/compiler/rustc_mir_transform/src/promote_consts.rs" "/checkout/compiler/rustc_mir_transform/src/coverage/tests.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Tue Mar 19 11:57:26 UTC 2024
  network time: Tue, 19 Mar 2024 11:57:26 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

Urgau commented:
Sorry for pings, messed-up the rebase.

@rustbot labels -WG-trait-system-refactor

Urgau committed:
Replace `RemapFileNameExt::for_codegen` with explicit calls

Urgau committed:
Simplify trim-paths feature by merging all debuginfo options together

Urgau committed:
Make local_crate_source_file return a RealFileName

so it can be remapped (or not) by callers

michaelwoerister commented on compiler/rustc_session/src/session.rs:
```diff
@@ -250,14 +251,9 @@ impl Session {
-            Some(path.to_path_buf())
-        }
+        Some(RealFileName::LocalPath(path.to_path_buf()))
     }
```
The method needs to apply the remapping. Otherwise `.for_scope` won't be able to return the remapped version because on the local file system path is available in the `RealFileName` value.

I suggest adding a helper method to [`FilePathMapping`](https://github.com/rust-lang/rust/blob/47dd709bedda8127e8daec33327e0a9d0cdae845/compiler/rustc_span/src/source_map.rs#L1047) to make creating `RealFileName`s easy:
```rust
pub fn to_real_filename<'a>(&'a self, local_path: impl Into<Cow<'a, Path>>) -> RealFileName {
    let local_path = local_path.into();
    if let (remapped_path, true) = self.map_prefix(&*local_path) {
        RealFileName::Remapped {
            virtual_name: remapped_path.into_owned(),
            local_path: Some(local_path.into_owned()),
        }
    } else {
        RealFileName::LocalPath(local_path.into_owned())
    }
}
```
Then you can implement `local_crate_source_file` as:
```rust
pub fn local_crate_source_file(&self) -> Option<RealFileName> {
    Some(self.source_map().path_mapping().to_real_filename(self.io.input.opt_path()?))
}
```

The `FilePathMapping::to_real_filename` helper can then also be used to create `Session::working_dir`.

Urgau replied:
Done and done.

*Btw, thanks you for the help and sorry this extra unnecessary work, I should have done the remapping from the start; I really shouldn't do 3 things at the same time 😅.*

michaelwoerister replied:
> I really shouldn't do 3 things at the same time 😅

I can relate to that one 😉

michaelwoerister commented on compiler/rustc_codegen_cranelift/src/debuginfo/mod.rs:
```diff
@@ -77,9 +77,16 @@ impl DebugContext {
                 FileNameDisplayPreference::Local
             })
             .into_owned();
```
Since `working_dir` is a  `RealFileName` we can do:
```rust
let comp_dir = tcx
    .sess
    .opts
    .working_dir
    .for_scope(tcx.sess, RemapPathScopeComponents::DEBUGINFO)
    .to_string_lossy();
```

michaelwoerister commented on compiler/rustc_codegen_cranelift/src/debuginfo/mod.rs:
```diff
@@ -75,9 +77,16 @@ impl DebugContext {
+                    } else {
+                        FileNameDisplayPreference::Local
+                    })
+                    .into_owned();
```
This can also be:
```rust
path
  .for_scope(tcx.sess, RemapPathScopeComponents::DEBUGINFO)
  .to_string_lossy();
```

michaelwoerister commented on compiler/rustc_codegen_llvm/src/back/write.rs:
```diff
@@ -257,13 +258,13 @@ pub fn target_machine_factory(
+            let path = if should_prefer_remapped_paths {
                 path_mapping.map_prefix(path).0
             } else {
                 path.into()
```
This could also make use of the `FilePathMapping::to_real_filename` helper proposed in the other comment.

Urgau replied:
Done.

Btw, this is the place where we cannot have a `Session` and therefor cannot use `for_scope`.

michaelwoerister commented on compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs:
```diff
@@ -875,7 +887,7 @@ pub fn build_compile_unit_di_node<'ll, 'tcx>(
+                if tcx.sess.should_prefer_remapped(RemapPathScopeComponents::DEBUGINFO) {
                     tcx.sess.source_map().path_mapping().map_prefix(f).0
                 } else {
                     f.into()
```
This could be expressed as
```
tcx
 .sess
 .source_map()
 .path_mapping()
 .to_real_filename(f)
 .for_scope(RemapPathScopeComponents::DEBUGINFO)
 .to_string_lossy()
```

michaelwoerister reviewed (commented):
I left a few comments below. In general, I think we should avoid duplicating the remapping logic at usage sites and go through `RealFileName` instead. Maybe we can even remove `Session::should_prefer_remapped` entirely.

Urgau committed:
Introduce `FileNameMapping::to_real_filename` and use it everywhere

Urgau commented:
Addressed the review comments, expect for 

> Maybe we can even remove Session::should_prefer_remapped entirely`

which sadly isn't possible as-is because both codegen (llvm and cranelift) are unable to hold a `Session` in their respective state struct.

michaelwoerister commented:
> which sadly isn't possible as-is because both codegen (llvm and cranelift) are unable to hold a Session in their respective state struct.

We can probably fix that by doing some refactoring. But that's not so urgent.

I think this PR looks good now. Thanks for the work you've put into it! I'll approve it some time next week so that folks on the tracking issue have a chance to voice concerns about the change (see https://github.com/rust-lang/rust/issues/111540#issuecomment-2014748609).

Urgau committed:
Replace Session should_remap_filepaths with filename_display_preference

Urgau commented:
Regarding the removal of `Session::should_prefer_remapped`, I think I found a pretty elegant solution. I've removed the function and replace-it with `Session::filename_display_preference` which returns a [`FileNameDisplayPreference`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/enum.FileNameDisplayPreference.html) enum which we then use in combination with [`display`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/enum.FileName.html#method.display) and [`to_string_display`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/enum.RealFileName.html#method.to_string_lossy). I've included this change in the last commit.

weihanglo commented on src/doc/unstable-book/src/compiler-flags/remap-path-scope.md:
```diff
@@ -10,10 +10,8 @@ This flag accepts a comma-separated list of values and may be specified multiple
-- `split-debuginfo-path` - apply remappings to the paths pointing to split debug information files. Does nothing when these files are not generated.
-- `object` - an alias for `macro,unsplit-debuginfo,split-debuginfo-path`. This ensures all paths in compiled executables or libraries are remapped, but not elsewhere.
+- `debuginfo` - apply remappings to debug informations
+- `object` - an alias for `macro,debuginfo`. This ensures all paths in compiled executables or libraries are remapped, but not elsewhere.
```
Is it possible that in the future `object` evolves beyond an alias for `macro,debuginfo`? I may suggest this:

```suggestion
- `object` - This ensures all paths in compiled executables or libraries are remapped, but not elsewhere. Currently an alias for `macro,debuginfo`.
```

michaelwoerister replied:
I think that's a good suggestion.

Urgau replied:
Agree, changed.

weihanglo reviewed (commented):
Looks nice. Thank you all for putting such much efforts on it. And sorry I didn't put much attention in the past few weeks 😞.

michaelwoerister commented:
> Regarding the removal of `Session::should_prefer_remapped`, I think I found a pretty elegant solution. I've removed the function and replace-it with `Session::filename_display_preference` which returns a [`FileNameDisplayPreference`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/enum.FileNameDisplayPreference.html) enum which we then use in combination with [`display`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/enum.FileName.html#method.display) and [`to_string_display`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/enum.RealFileName.html#method.to_string_lossy). I've included this change in the last commit.

That is a nice solution. Thanks!

michaelwoerister commented on compiler/rustc_session/src/session.rs:
```diff
@@ -886,38 +882,15 @@ impl Session {
+    pub fn filename_display_preference(
+        &self,
+        scope: RemapPathScopeComponents,
+    ) -> FileNameDisplayPreference {
```
Can you also add the same assertion we have in `for_scope`:
```rust
assert!(
    scope.bits().count_ones() == 1,
    "one and only one scope should be passed to for_scope"
);
```

Urgau replied:
Sure, done.

michaelwoerister reviewed (commented):
I found one thing that I'd like to be changed. With that fixed, I'm happy to approve the PR.

michaelwoerister commented:
Thanks, @Urgau!

@bors r+

bors commented:
:pushpin: Commit f2a2c4c9ef8443329cc0d0a0a7dd5fdbf364b001 has been approved by `michaelwoerister`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=michaelwoerister f2a2c4c9ef8443329cc0d0a0a7dd5fdbf364b001 -->
<!-- homu: {"type":"Approved","sha":"f2a2c4c9ef8443329cc0d0a0a7dd5fdbf364b001","approver":"michaelwoerister","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #123166) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Urgau commented:
Rebased *(and fix some typos)*.

@rustbot ready

michaelwoerister commented:
Thanks!

@bors r+

bors commented:
:pushpin: Commit fefb8f1f9cb830e1ca0b0ba449db7c8e7d4ff7ba has been approved by `michaelwoerister`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=michaelwoerister fefb8f1f9cb830e1ca0b0ba449db7c8e7d4ff7ba -->
<!-- homu: {"type":"Approved","sha":"fefb8f1f9cb830e1ca0b0ba449db7c8e7d4ff7ba","approver":"michaelwoerister","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit fefb8f1f9cb830e1ca0b0ba449db7c8e7d4ff7ba with merge 685927aae69657b46323cffbeb0062835bd7fa2b...
<!-- homu: {"type":"BuildStarted","head_sha":"fefb8f1f9cb830e1ca0b0ba449db7c8e7d4ff7ba","merge_sha":"685927aae69657b46323cffbeb0062835bd7fa2b"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8481740909/job/23242961129)
Approved by: michaelwoerister
Pushing 685927aae69657b46323cffbeb0062835bd7fa2b to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"michaelwoerister","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8481740909/job/23242961129"},"merge_sha":"685927aae69657b46323cffbeb0062835bd7fa2b"} -->

rust-timer commented:
Finished benchmarking commit (685927aae69657b46323cffbeb0062835bd7fa2b): [comparison URL](https://perf.rust-lang.org/compare.html?start=45796d1c24445b298567752519471cef2cff3298&end=685927aae69657b46323cffbeb0062835bd7fa2b&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=45796d1c24445b298567752519471cef2cff3298&end=685927aae69657b46323cffbeb0062835bd7fa2b&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.2% | [0.2%, 0.2%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=45796d1c24445b298567752519471cef2cff3298&end=685927aae69657b46323cffbeb0062835bd7fa2b&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=45796d1c24445b298567752519471cef2cff3298&end=685927aae69657b46323cffbeb0062835bd7fa2b&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=45796d1c24445b298567752519471cef2cff3298&end=685927aae69657b46323cffbeb0062835bd7fa2b&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 668.043s -> 667.959s (-0.01%)
**Artifact size**: 315.85 MiB -> 315.84 MiB (-0.00%)
